### PR TITLE
[SPARK-42169] Implement code generation for to_csv function (StructsToCsv)

### PR DESCRIPTION
**Pull request by NarekDW**
### What changes were proposed in this pull request?
This PR enhances `StructsToCsv` class with `doGenCode` function instead of extending it from `CodegenFallback` trait (performance improvement).

### Why are the changes needed?
It will improve performance.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
by existing tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala [hunk 1]
```diff
@@ -245,8 +245,8 @@ case class StructsToCsv(
      options: Map[String, String],
      child: Expression,
      timeZoneId: Option[String] = None)
-  extends UnaryExpression with TimeZoneAwareExpression with CodegenFallback with ExpectsInputTypes
-    with NullIntolerant {
+  extends UnaryExpression with TimeZoneAwareExpression with ExpectsInputTypes
+    with NullIntolerant with Serializable {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala:249]: Case class is serializable so won't need `with Serializable `.

**Comment by HyukjinKwon** [comment]
Hm, is the issue from non-codegen part from `Rand`?

**Comment by HyukjinKwon** [comment]
Would be easier to review if you describe how it happens and how this PR fixes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala (hunk 1 above)
Comment (NarekDW) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala]: sure, fixed

**Comment by NarekDW** [comment]
> 

Hi @HyukjinKwon I've added 2 sections in the header description - **Debugging the root cause** and **How this PR changes such a behaviour** with my point of view, how it happens and how this PR could improve the behaviour. 
Could you review, please.

**Comment by HyukjinKwon** [comment]
cc @cloud-fan FYI

**Comment by gboo-infa** [comment]
While this change may be good to have for its own reasons, it doesn't really address the problem in the JIRA. The problem is that CodegenFallback is incompatible with nondeterministic child nodes; the problem is not that to_csv happens to be CodegenFallback. Other nodes are also CodegenFallback, and future nodes will be as well, and this change doesn't address that...

**Comment by NarekDW** [comment]
> While this change may be good to have for its own reasons, it doesn't really address the problem in the JIRA. The problem is that CodegenFallback is incompatible with nondeterministic child nodes; the problem is not that to_csv happens to be CodegenFallback. Other nodes are also CodegenFallback, and future nodes will be as well, and this change doesn't address that...

Yep, good point. This is just a partial solution, for `to_csv` function in particular.
But I don't know, should we rename this PR or create another task/bug in JIRA...

**Comment by NarekDW** [comment]
@HyukjinKwon any suggestion on this?
Should I keep this PR open?

**Comment by gboo-infa** [comment]
Oops sorry, I was hoping someone with authority or at least familiarity with the Spark committer culture would answer your question. I don't really have either of those things so I can't really comment on whether Spark JIRAs are ok to use for related issues...

My only concern was that someone would mistake your PR as a fix for the underlying issue and mark the JIRA as resolved, and I didn't want to cause the confusion of reopening the JIRA or submitting an identical one with a different example operation.

Personally, I think if you've already gone to the trouble to create the PR, and it doesn't break anything, then maybe it's not so bad to just check it in regardless of what JIRA it uses... It'll probably make it harder for people to make sense of the release notes, so a few extra minutes from you might save hours of collective time for the rest of the world, but I dunno, up to you. :/

**Comment by cloud-fan** [comment]
I'm fixing the root cause at https://github.com/apache/spark/pull/39248

**Comment by cloud-fan** [comment]
Can we refine the PR description? It still good to have codegen for `to_csv`, but it's only for performance, as the correctness issue has been fixed by #39248

**Comment by NarekDW** [comment]
> Can we refine the PR description? It still good to have codegen for `to_csv`, but it's only for performance, as the correctness issue has been fixed by #39248

@cloud-fan Sure, I've renamed this PR (also changed the description). Unfortunately, I don't have an account in the JIRA, and I can't create a separate JIRA task for this.  
Also, I left a test case in this PR:
```scala
test("SPARK-41049: make to_csv function deterministic") {
...
```
 I think it might be useful, let me know if it is redundant.

File: sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala [hunk 2]
```diff
@@ -574,4 +574,36 @@ class CsvFunctionsSuite extends QueryTest with SharedSparkSession {
       $"csv", schema_of_csv("1,2\n2"), Map.empty[String, String].asJava))
     checkAnswer(actual, Row(Row(1, "2\n2")))
   }
+
+  test("SPARK-41049: make to_csv function deterministic") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:578]: I think this test is already covered by https://github.com/apache/spark/pull/39248/files#diff-0b6750da4fdbe797612d14fa705138892221bca66e748241ff390b58ab4e5352R3571

**Comment by cloud-fan** [comment]
> Unfortunately, I don't have an account in the JIRA, and I can't create a separate JIRA task for this.

Can you sign up and create an account? The JIRA system allows anyone to register.

**Comment by NarekDW** [comment]
> > Unfortunately, I don't have an account in the JIRA, and I can't create a separate JIRA task for this.
> 
> Can you sign up and create an account? The JIRA system allows anyone to register.

Unfortunately, it is disabled
image

I mailed to private@spark.apache.org to get an account created, about 2 weeks ago, but didn't get any response...
If you can suggest anybody, whom I can address with such a question, it will be great.

File: sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala (hunk 2 above)
Comment (NarekDW) [reply sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala]: removed

**Comment by NarekDW** [comment]
@cloud-fan finally I've got my JIRA account and created a separate ticket for this.
I've closed this PR as I've renamed my branch in accordance with new JIRA ticket.
Please check the PR here - https://github.com/apache/spark/pull/39719

**Comment by weicm** [comment]
> @cloud-fan finally I've got my JIRA account and created a separate ticket for this. I've closed this PR as I've renamed my branch in accordance with new JIRA ticket. Please check the new PR here - #39719

Are you getting a JIRA account by sending an email to private@spark.apache.org?

**Pull request closed without merging**